sips/sip-021/sip-021-nakamoto.md

```diff
@@ -0,0 +1,1051 @@
+`SP00000000000000DSQJTCSE63RMXHDP`. Then, his 1,000 STX are tabulated as "no."
+
+# Reference Implementation
+
+The reference implementation can be found at
+https://github.com/stacks-network/stacks-blockchain.
+
+## Stacker Responsibilities
+
+The act of Stacking requires the Stacker to be online 24/7 to sign blocks.  To
+facilitate this, the implementation comes with a Stacker signer daemon, which
+runs as an event observer to the Stacks node.
+
+The Stacker signer daemon receives notifications from the Stacks node when a new
+block arrives.  On receipt of the block, the daemon instigates a WSTS signing
+round with other signer daemons to generate an aggregate Schnorr signature.
+
+The Stacker signer daemons communicate with one another through a network
+overlay within the Stacks peer-to-peer network called a _StackerDB_.  The
+StackerDB system allows nodes to replicate an array of fixed-sized chunks of
+arbitrary data, which must be signed by principals identified by a smart
+contract.
+
+## StackerDB
+
+StackerDB is a feature that will ship prior to this SIP's activation.  It allows
+users to store data within the Stacks peer-to-peer network by means of a
+specially-crafted smart contract, and a connected overlay network.  The smart
+contract describes the parameters of the data (e.g. who can write to it; how
+much data can be stored; and so on); the data itself is stored off-chain.  A
+StackerDB-aware node maintains connections to other StackerDB-aware nodes who
+replicate the same StackerDBs as itself.
+
+The StackerDB data schema is an array of fixed-sized chunks.  Each chunk has an
+slot index, a monotonically-increasing version, and a signer (e.g. a Stacks
+address).  A user writes a chunk by POST-ing new chunk data for the slot, as
+well as a new version number and a signature over both the data and the version.
+If the chunk is newer than the chunk already stored (as identified by version
+number), then the node stores the chunk and replicates it to other nodes
+subscribed to the same StackerDB instance.
+
+Stacks nodes announce which StackerDB replicas they subscribe to when they
+handshake with one another.  If both the handshaker and its peer support the
+StackerDB protocol, they exchange the list of replicas that they maintain.  In
+doing so, StackerDB-aware nodes eventually learn about all reachable nodes'
+StackerDB replicas as they walk the peer graph.
+
+StackerDB-aware nodes set up overlay networks on top of the Stacks peer-to-peer
+network to replicate StackerDB chunks.  Nodes that replicate a particular
+StackerDB will periodically exchange version vectors for the list of chunks in
+the replica.  If one node discovers that another node has a newer version of the
+chunk, it will download it and forward it to other StackerDB-aware nodes in the
+overlay that also need it.  In doing so, every StackerDB-aware node in the
+overlay eventually receives an up-to-date replica of the StackerDB chunks.
```

Comment:
Might be worthwhile to mention how long this data is stored. I think its for at most 2 reward cycles right?

Reply:
It also might be worthwhile to link to the stackerdb code docs if we have any (pretty sure we do. Could swear Jude sent me a link to them before)

Reply:
It's indefinite -- as long as the slot exists across reconfiguration, it will continue to be preserved